crates/rpc/rpc/src/eth/api/traits/state.rs

```diff
@@ -0,0 +1,173 @@
+            Ok(this
+                .state_at_block_id_or_latest(block_id)?
+                .account_code(address)?
+                .unwrap_or_default()
+                .original_bytes())
+        })
+    }
+
+    /// Returns balance of given account, at given blocknumber.
+    fn balance(
+        &self,
+        address: Address,
+        block_id: Option<BlockId>,
+    ) -> impl Future<Output = EthResult<U256>> + Send {
+        self.spawn_blocking_io(move |this| {
+            Ok(this
+                .state_at_block_id_or_latest(block_id)?
+                .account_balance(address)?
+                .unwrap_or_default())
+        })
+    }
+
+    /// Returns values stored of given account, at given blocknumber.
+    fn storage_at(
+        &self,
+        address: Address,
+        index: JsonStorageKey,
+        block_id: Option<BlockId>,
+    ) -> impl Future<Output = EthResult<B256>> + Send {
+        self.spawn_blocking_io(move |this| {
+            Ok(B256::new(
+                this.state_at_block_id_or_latest(block_id)?
+                    .storage(address, index.0)?
+                    .unwrap_or_default()
+                    .to_be_bytes(),
+            ))
+        })
+    }
+
+    /// Returns values stored of given account, with Merkle-proof, at given blocknumber.
+    fn get_proof(
+        &self,
+        address: Address,
+        keys: Vec<JsonStorageKey>,
+        block_id: Option<BlockId>,
+    ) -> EthResult<impl Future<Output = EthResult<EIP1186AccountProofResponse>> + Send>
+    where
+        Self: EthApiSpec,
+    {
+        let chain_info = self.chain_info()?;
+        let block_id = block_id.unwrap_or_default();
+
+        // if we are trying to create a proof for the latest block, but have a BlockId as input
+        // that is not BlockNumberOrTag::Latest, then we need to figure out whether or not the
+        // BlockId corresponds to the latest block
+        let is_latest_block = match block_id {
+            BlockId::Number(BlockNumberOrTag::Number(num)) => num == chain_info.best_number,
+            BlockId::Hash(hash) => hash == chain_info.best_hash.into(),
+            BlockId::Number(BlockNumberOrTag::Latest) => true,
+            _ => false,
+        };
+
+        // TODO: remove when HistoricalStateProviderRef::proof is implemented
+        if !is_latest_block {
+            return Err(EthApiError::InvalidBlockRange)
+        }
+
+        Ok(self.spawn_tracing(move |this| {
+            let state = this.state_at_block_id(block_id)?;
+            let storage_keys = keys.iter().map(|key| key.0).collect::<Vec<_>>();
+            let proof = state.proof(address, &storage_keys)?;
+            Ok(from_primitive_account_proof(proof))
+        }))
+    }
+}
+
+/// Loads state from database.
+pub trait LoadState {
+    /// Returns a handle for reading state from database.
+    ///
+    /// Data access in default trait method implementations.
+    fn provider(&self) -> &impl StateProviderFactory;
```

Comment:
is auto_impl for &, so this should work
```suggestion
    fn provider(&self) -> impl StateProviderFactory;
```